src/NServiceBus.Extensions.Hosting/NServiceBusHostedService.cs

```diff
@@ -4,29 +4,44 @@
     using System.Threading;
     using System.Threading.Tasks;
     using Microsoft.Extensions.Hosting;
-    using NServiceBus;
 
     class NServiceBusHostedService : IHostedService
     {
-        public NServiceBusHostedService(IStartableEndpointWithExternallyManagedContainer startableEndpoint, IServiceProvider serviceProvider)
+        public NServiceBusHostedService(IStartableEndpointWithExternallyManagedContainer startableEndpoint)
         {
             this.startableEndpoint = startableEndpoint;
-            this.serviceProvider = serviceProvider;
         }
 
+        public Task<IEndpointInstance> Endpoint => endpointTcs.Task;
+
         public async Task StartAsync(CancellationToken cancellationToken)
         {
-            endpoint = await startableEndpoint.Start(new ServiceProviderAdapter(serviceProvider))
-                .ConfigureAwait(false);
+            try
+            {
+                var endpoint = await startableEndpoint.Start(new ServiceProviderAdapter(serviceProvider))
+                    .ConfigureAwait(false);
+                endpointTcs.SetResult(endpoint);
+            }
+            catch (Exception e)
+            {
+                endpointTcs.SetException(e);
+                throw;
+            }
+        }
+
+        public async Task StopAsync(CancellationToken cancellationToken)
+        {
+            var endpoint = await endpointTcs.Task.ConfigureAwait(false);
```

Comment:
If endpoint start bombs `SetException` is called if the hosted service thingy still calls `StopAsync` then this line would throw again the same exception. Not sure what the hosted service middleware does and how it behaves. Just mentioning this case here

Reply:
from what I saw so far, the infrastructure doesn't call stop on services which threw on start.